sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c

```diff
@@ -0,0 +1,344 @@
+            mlme_confirm->mib.activation = mac->mlme.activation;
+            break;
+        default:
+            mlme_confirm->status = -EINVAL;
+            break;
+    }
+}
+
+void gnrc_lorawan_mlme_request(gnrc_lorawan_t *mac, const mlme_request_t *mlme_request,
+                               mlme_confirm_t *mlme_confirm)
+{
+    switch (mlme_request->type) {
+        case MLME_JOIN:
+            if(mac->mlme.activation != MLME_ACTIVATION_NONE) {
+                mlme_confirm->status = -EINVAL;
+                break;
+            }
+            if (!gnrc_lorawan_mac_acquire(mac)) {
+                mlme_confirm->status = -EBUSY;
+                break;
+            }
+
+            if (mac->mlme.backoff_budget < 0) {
+                mlme_confirm->status = -EDQUOT;
+                break;
+            }
+            memcpy(mac->appskey, mlme_request->join.appkey, LORAMAC_APPKEY_LEN);
+            mlme_confirm->status = gnrc_lorawan_send_join_request(mac, mlme_request->join.deveui,
+                                                                  mlme_request->join.appeui, mlme_request->join.appkey, mlme_request->join.dr);
+            break;
+        case MLME_LINK_CHECK:
+            mac->mlme.pending_mlme_opts |= GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_DEFERRED;
+            break;
+        case MLME_SET:
+            _mlme_set(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_GET:
+            _mlme_get(mac, mlme_request, mlme_confirm);
+            break;
+        case MLME_RESET:
+            gnrc_lorawan_reset(mac);
+            mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+            break;
+        default:
+            break;
+    }
+}
+
+int _fopts_mlme_link_check_req(lorawan_buffer_t *buf)
+{
+    if (buf) {
+        assert(buf->index + GNRC_LORAWAN_CID_SIZE <= buf->size);
+        buf->data[buf->index++] = GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS;
+    }
+
+    return GNRC_LORAWAN_CID_SIZE;
+}
+
+static int _mlme_link_check_ans(gnrc_lorawan_t *mac, lorawan_buffer_t *fopt)
+{
+    if (fopt->index + GNRC_LORAWAN_FOPT_LINK_ANS_SIZE > fopt->size) {
+        return -EINVAL;
+    }
+    fopt->index++;
+
+    mlme_confirm_t *mlme_confirm = gnrc_lorawan_mlme_allocate(mac);
+    mlme_confirm->link_req.margin = fopt->data[fopt->index++];
+    mlme_confirm->link_req.num_gateways = fopt->data[fopt->index++];
+
+    mlme_confirm->type = MLME_LINK_CHECK;
+    mlme_confirm->status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
+    mac->netdev.event_callback(&mac->netdev, NETDEV_EVENT_MLME_CONFIRM);
+
+    mac->mlme.pending_mlme_opts &= ~GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+
+    return 0;
+}
+
+void gnrc_lorawan_process_fopts(gnrc_lorawan_t *mac, uint8_t *fopts, size_t size)
+{
+    if (!fopts || !size) {
+        return;
+    }
+
+    lorawan_buffer_t buf;
+    _buffer_reset(&buf, fopts, size);
+
+    while (buf.index < buf.size) {
```

Comment:
I find this a bit hard to understand, bc just looking at this it seems to be a infinite loop. 
Actually its not bc `_mlme_link_check_ans` changes an internal value of `buf`. 

I would rewrite this to something like:

```
int ret = 0;
while ((buf.index < buf.size) && (ret >= 0)) {
    buf.index += ret;
    switch (*(buf.data)) {
        case GNRC_LORAWAN_CID_LINK_CHECK_REQ_ANS:
            ret = _mlme_link_check_ans(mac, &buf);
            break;
        default:
            ret = -1;
     }
}
```

then rewrite `_mlme_link_check_ans ` to return `3` as how far the index needs to me moved forward.

What do you think?

Reply:
I replaced this whole snippet to keep check of fopts flow (malformed fopts frame, very unlikely but not out of the specification...)

Besides saving some ROM, the link check command still works:
```
2019-11-21 15:03:27,349 # Iface  4  HWaddr: 26:01:29:77  Frequency: 869524963Hz  BW: 125kHz  SF: 9  CR: 4/5  Link: up 
2019-11-21 15:03:27,356 #            TX-Power: 14dBm  State: SLEEP  Demod margin.: 2  Num gateways.: 25 
2019-11-21 15:03:27,359 #           IQ_INVERT 
2019-11-21 15:03:27,362 #           RX_SINGLE OTAA L2-PDU:255 
```

Reply:
@smlng are you OK with the proposed solution? If so, all comments are addressed